cns/refresh/fetcher.go

```diff
@@ -0,0 +1,108 @@
+package refresh
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	"github.com/Azure/azure-container-networking/cns/logger"
+)
+
+const (
+	DefaultMinInterval = 4 * time.Second
+	DefaultMaxInterval = 1024 * time.Second
+)
+
+// Fetcher fetches data at regular intervals. The interval will vary within the range of minInterval and
+// maxInterval. When no diff is observed after a fetch, the interval doubles (subject to the maximum interval).
+// When a diff is observed, the interval resets to the minimum.
+
+type Fetcher[T any] struct {
+	fetchFunc       func(context.Context) (T, error)
+	cache           T
+	minInterval     time.Duration
+	maxInterval     time.Duration
+	currentInterval time.Duration
+	ticker          TickProvider
+	consumeFunc     func(T) error
+}
+
+// NewFetcher creates a new Fetcher. If minInterval is 0, it will default to 4 seconds.
+func NewFetcher[T any](fetchFunc func(context.Context) (T, error), minInterval, maxInterval time.Duration, consumeFunc func(T) error) *Fetcher[T] {
+	if minInterval == 0 {
+		minInterval = DefaultMinInterval
+	}
+
+	if maxInterval == 0 {
+		maxInterval = DefaultMaxInterval
+	}
+
+	maxInterval = max(minInterval, maxInterval)
+
+	return &Fetcher[T]{
+		fetchFunc:       fetchFunc,
+		minInterval:     minInterval,
+		maxInterval:     maxInterval,
+		currentInterval: minInterval,
+		consumeFunc:     consumeFunc,
+	}
+}
+
+func (f *Fetcher[T]) Start(ctx context.Context) {
+	go func() {
+		// do an initial fetch
+		res, err := f.fetchFunc(ctx)
+		if err != nil {
+			logger.Printf("Error refreshing secondary IPs: %v", err)
+		}
+
+		f.cache = res
+		if f.consumeFunc != nil {
+			if err := f.consumeFunc(res); err != nil {
+				logger.Errorf("Error consuming data: %v", err)
+			}
+		}
+
+		if f.ticker == nil {
+			f.ticker = NewTimedTickProvider(f.currentInterval)
+		}
+
+		defer f.ticker.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Printf("Fetcher stopped")
+				return
+			case <-f.ticker.C():
+				result, err := f.fetchFunc(ctx)
+				if err != nil {
+					logger.Errorf("Error fetching data: %v", err)
+				} else {
+					if reflect.DeepEqual(result, f.cache) {
```

Comment:
Never use `reflect.DeepEqual`, use `github.com/google/go-cmp/cmp.Equal` instead (it's a drop-in replacement that is safer).

Reply:
consider making the type implement `Comparable` instead of either of these

Reply:
@rbtr As far as I could tell, comparable is just a characteristic of the struct, not an interface I could implement (https://go.dev/ref/spec#Comparison_operators). Could you please provide more info?

Reply:
@rbtr I could add the constraint, but it seems struct types either satisfies the constraint or not based on their members. Specifically, the struct we are currently passing wouldn't satisfy the comparable constraint due to having slice members. Generally that would restrict potential structs to much I feel.

Reply:
> @rbtr I could add the constraint, but it seems struct types either satisfy the constraint or not based on their members (we can't make a struct satisfy it, without editing the members). Specifically, the struct we are currently passing wouldn't satisfy the comparable constraint due to having slice members. Generally that would restrict potential structs to much I feel.

Reply:
So, @rbtr and I talked offline. I had tried the Comparable route and it was not fruitful for the reasons you mention @santhoshmprabhu . Also, my previous thought about using `cmp.Equal` was incorrect since that's entirely useful only for tests (usually people reach for `reflect.DeepEqual` in tests, where that advice usually applies). In this case, instead of parameterizing the type on `any`, parameterize it on a generic interface:

```go
type Equalable[T any] interface{
       Equal(other T) bool
}

type Fetcher[T Equalable[T]] struct {
       // snip
}
```

Then anyone using this is forced to implement `Equal(other T) bool`, which can provide equality guarantees through unit tests... then we can just call that method.

Reply:
I had done this before I saw @timraymond 's comment, so the names are a little different. Please take a look.
cc: @rbtr

Reply:
I am somewhat concerned about the amount of copying that the collection of Equal methods is adding, so if you folks think we need the function by value only for Interface and the rest can take pointers, that would be more efficient.